src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpSimplifyTypeNamesDiagnosticAnalyzer.cs

```diff
@@ -26,60 +29,39 @@ internal sealed class CSharpSimplifyTypeNamesDiagnosticAnalyzer
                 SyntaxKind.SimpleMemberAccessExpression,
                 SyntaxKind.QualifiedCref);
 
-        public CSharpSimplifyTypeNamesDiagnosticAnalyzer()
-            : base(s_kindsOfInterest)
+        public override void Initialize(AnalysisContext context)
         {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(AnalyzeCompilation);
         }
 
-        protected override void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        private void AnalyzeCompilation(CompilationStartAnalysisContext context)
         {
-            if (context.Node.Ancestors(ascendOutOfTrivia: false).Any(n => !n.IsKind(SyntaxKind.QualifiedCref) && s_kindsOfInterest.Contains(n.Kind())))
-            {
-                // Bail out early because we have already simplified an ancestor of this node (except in the QualifiedCref case).
-                // We need to keep going in case this node is under a QualifiedCref because it is possible to have multiple simplifications within the same QualifiedCref.
-                // For example, consider <see cref="A.M(Nullable{int})"/>. The 'A.M(Nullable{int})' here is represented by a single QualifiedCref node in the syntax tree.
-                // It is possible to have a simplification to remove the 'A.' qualification for the QualifiedCref itself as well as another simplification to change 'Nullable{int}'
-                // to 'int?' in the GenericName for the 'Nullable{T}' that is nested inside this QualifiedCref. We need to keep going so that the latter simplification can be
-                // made available.
-                return;
-            }
+            var compilationTypeNames = new HashSet<string>();
+            compilationTypeNames.AddAll(context.Compilation.Assembly.TypeNames);
 
-            Diagnostic diagnostic;
-            var options = context.Options;
-            var cancellationToken = context.CancellationToken;
-            bool descendIntoChildren(SyntaxNode n)
-            {
-                // Don't examine crefs here, we'll process it in the loop below that
-                // descends into trivia.
-                if (IsCrefCandidate(n))
-                    return true;
+            context.RegisterSemanticModelAction(c => AnalyzeSemanticModel(c, compilationTypeNames));
```

Comment:
Can we not use `RegisterSemanticModelAction`? This API does not have optimal performance characteristics due to the fact that it does not share semantic model with compiler's method body binding. It will *always* use a different semantic model instance then the one used by the compiler for method body binding. I think this analysis should be written as a hybrid syntax node + semantic analysis, where syntax node actions are used for analysis outside executable code blocks and IOperation callbacks are used for semantic analysis inside executable code blocks.

Reply:
>  It will always use a different semantic model instance then the one used by the compiler for method body binding.

That seems... terrible.  Why does it do that?

Reply:
@mavasani if i register for CompilationUnitSyntax, will i get called back with teh same SemanticModel the compiler uses?

Reply:
Nope. I believe the primary reason is compiler creates a different method body semantic model while binding each method body, and uses it for IOperation callbacks and syntax node callbacks within that method body and then drops it on the floor. Compiler does not want to hold onto method body semantic model outside the method body binding. Sharing the same semantic model for any action scoped beyond the method body would need them to hold onto it longer.

Reply:
#wontfix

need the compiler to improve things here.